File: cleanup/test/correctness/ipm/ipm.jl
using MatrixMarket
using MatrixMarket2

include("dss.jl")

function speye_int32(m::Integer)
  rowval = [Int32(x) for x in [1:m;]]
  colptr = [Int32(x) for x in [rowval; m + 1]]
  nzval = ones(Float64, m)
  return SparseMatrixCSC(m, m, colptr, rowval, nzval)
end

# A*D*B, where D is a diagonal matrix
# Since it's implemented in CSR, we pass tranposes of A and B
# in CSC
function adb_inspect(AT::SparseMatrixCSC, BT::SparseMatrixCSC)
  csrA = SparseAccelerator.CreateCSR(AT)
  csrB = SparseAccelerator.CreateCSR(BT)

  # A*D*A' inspection
  csrADAT = ccall((:CSR_ADBInspect, LIB_CSR_PATH), Ptr{Void},
                  (Ptr{Void}, Ptr{Void}),
                  csrA, csrB)

  m = size(A, 1)
  rowptr = pointer_to_array(
    ccall((:CSR_GetRowPtr, LIB_CSR_PATH), Ptr{Cint},
          (Ptr{Void},),
          csrADAT),
    (m + 1,))
  nnz = rowptr[m + 1] - 1
  colidx = pointer_to_array(
    ccall((:CSR_GetColIdx, LIB_CSR_PATH), Ptr{Cint},
          (Ptr{Void},),
          csrADAT),
    (nnz,))
  values = pointer_to_array(
    ccall((:CSR_GetValues, LIB_CSR_PATH), Ptr{Cdouble},
          (Ptr{Void},),
          csrADAT),
    (nnz,))

  SparseAccelerator.DestroyCSR(csrA)
  SparseAccelerator.DestroyCSR(csrB)

  ADB = SparseMatrixCSC{Cdouble, Cint}(
    m, m, rowptr, colidx, values)
end

# A*D*B, where D is a diagonal matrix
function adb_execute!(ADB::SparseMatrixCSC, AT::SparseMatrixCSC, BT::SparseMatrixCSC, d::Vector)
  csrA = SparseAccelerator.CreateCSR(AT)
  csrB = SparseAccelerator.CreateCSR(BT)
  csrADB = SparseAccelerator.CreateCSR(ADB)

  ccall((:CSR_ADB, LIB_CSR_PATH), Void,
        (Ptr{Void}, Ptr{Void}, Ptr{Void}, Ptr{Cdouble}),
        csrADB, csrA, csrB, d)

  SparseAccelerator.DestroyCSR(csrA)
  SparseAccelerator.DestroyCSR(csrB)
  SparseAccelerator.DestroyCSR(csrADB)
end

# optimized implementation of interior-point method with inspector hoisted
function ipm(A, b, p, print = true) # A: constraint coefficients, b: constraint rhs, p: objective
  t0 = time()
  (m,n) = size(A)

# set initial point, based on largest element in (A,b,p)
  bigM = maximum(A)
  bigM = maximum([norm(b, Inf) norm(p, Inf) bigM])
  x = 100*bigM*ones(n)
  s = x
  y = zeros(m)

  bc = 1 + maximum([norm(b) norm(p)])

  relResidual = NaN
  iter=1

  blas1_time = 0.
  spgemm_time = 0.
  fact_time = 0.
  trslv_time = 0.
  spmv_time = 0.

  spmv_inspector_time = 0.
  spgemm_inspector_time = 0.
  fact_inspector_time = 0.

  D = speye_int32(n)

  spmv_inspector_time -= time()
  AT = A'
  spmv_inspector_time += time()

  spgemm_inspector_time -= time()
  ADAT = adb_inspect(AT, A)
  spgemm_inspector_time += time()

  fact_inspector_time -= time()

  dss_handle = dss_analyze(ADAT)
  dy = Array(Cdouble, m)

  fact_inspector_time += time()

  for iter=1:200

    # compute residuals
    spmv_time -= time()
    Rd = AT*y + s - p
    Rp = A*x - b
    spmv_time += time()

    blas1_time -= time()
    Rc = x.*s
    mu = mean(Rc)
    relResidual = norm([Rd; Rp; Rc])/bc
    blas1_time += time()

    if print
      #@printf "iter %2i: log10(mu) = %5.2f, resid = %9.2e, obj = %9.2e\n" iter log10(mu) relResidual (p'*x)[1][1]
    end
    if (relResidual <= 1e-7 && mu <= 1e-7) break; end

    blas1_time -= time()
    Rc = Rc - min(0.1, 100*mu)*mu

    # set up the scaling matrix, and form the coefficient matrix for
    # the linear system
    d = min(5.e+15, x./s)
    blas1_time += time()

    spgemm_time -= time()
    adb_execute!(ADAT, AT, A, d)
    spgemm_time += time()

    # use the form of the Cholesky routine "cholinc" that's best
    # suited to interior-point methods
    fact_time -= time()
    dss_factor(dss_handle, ADAT)
    fact_time += time()

    # set up the right-hand side
    blas1_time -= time()
    t1 = x.*Rd - Rc;
    blas1_time += time()

    spmv_time -= time()
    t2 = -(Rp + A*(t1./s));
    spmv_time += time()

    # solve it and recover the other step components
    trslv_time -= time()
    opt = MKL_DSS_DEFAULTS
    dss_solve!(dss_handle, t2, dy)
    trslv_time += time()

    spmv_time -= time()
    temp = AT*dy
    spmv_time += time()

    blas1_time -= time()
    dx = (x.*temp + t1)./s
    ds = -(s.*dx + Rc)./x

    tau = max(.9995, 1 - mu)
    ap = -1/minimum([dx./x; -1])
    ad = -1/minimum([ds./s; -1])
    ap = tau*ap
    ad = tau*ad
    x = x + ap*dx
    s = s + ad*ds
    y = y + ad*dy
    blas1_time += time()
  end

  blas1_time -= time()
  f = p'*x
  blas1_time += time()

  if print
    @printf "\nwith_inspector_hoisting_total_time = %f\n" time() - t0
    @printf "spgemm = %f fact = %f blas1 = %f trslv = %f spmv = %f\n" spgemm_time fact_time blas1_time trslv_time spmv_time
    @printf "spgemm_inspector = %f fact_inspector_time = %f spmv_inspector = %f\n" spgemm_inspector_time fact_inspector_time spmv_inspector_time
    @printf "iter %2i, resid = %9.2e, objval = %e\n" iter relResidual f[1][1]
  end

  x
end

# unoptimized implementation of interior-point method w/o inspector hoisting
function ipm_unopt(A, b, p, print = true) # A: constraint coefficients, b: constraint rhs, p: objective
  t0 = time()
  (m,n) = size(A)

# set initial point, based on largest element in (A,b,p)
  bigM = maximum(A)
  bigM = maximum([norm(b, Inf) norm(p, Inf) bigM])
  x = 100*bigM*ones(n)
  s = x
  y = zeros(m)

  bc = 1 + maximum([norm(b) norm(p)])

  relResidual = NaN
  iter=1

  blas1_time = 0.
  spgemm_time = 0.
  fact_time = 0.
  trslv_time = 0.
  spmv_time = 0.

  D = speye_int32(n)
  dy = Array(Cdouble, m)

  for iter=1:200

    # compute residuals
    spmv_time -= time()
    Rd = A'*y + s - p
    Rp = A*x - b
    spmv_time += time()

    blas1_time -= time()
    Rc = x.*s
    mu = mean(Rc)
    relResidual = norm([Rd; Rp; Rc])/bc
    blas1_time += time()

    if print
      #@printf "iter %2i: log10(mu) = %5.2f, resid = %9.2e, obj = %9.2e\n" iter log10(mu) relResidual (p'*x)[1][1]
    end
    if (relResidual <= 1e-7 && mu <= 1e-7) break; end

    blas1_time -= time()
    Rc = Rc - min(0.1, 100*mu)*mu

    # set up the scaling matrix, and form the coefficient matrix for
    # the linear system
    d = min(5.e+15, x./s)
    blas1_time += time()

    spgemm_time -= time()
    ADAT = adb_inspect(A', A)
    adb_execute!(ADAT, A', A, d)
    spgemm_time += time()

    # use the form of the Cholesky routine "cholinc" that's best
    # suited to interior-point methods
    fact_time -= time()
    dss_handle = dss_analyze(ADAT)
    dss_factor(dss_handle, ADAT)
    fact_time += time()

    # set up the right-hand side
    blas1_time -= time()
    t1 = x.*Rd - Rc;
    blas1_time += time()

    spmv_time -= time()
    t2 = -(Rp + A*(t1./s));
    spmv_time += time()

    # solve it and recover the other step components
    trslv_time -= time()
    opt = MKL_DSS_DEFAULTS
    dss_solve!(dss_handle, t2, dy)
    trslv_time += time()

    spmv_time -= time()
    temp = A'*dy
    spmv_time += time()

    blas1_time -= time()
    dx = (x.*temp + t1)./s
    ds = -(s.*dx + Rc)./x

    tau = max(.9995, 1 - mu)
    ap = -1/minimum([dx./x; -1])
    ad = -1/minimum([ds./s; -1])
    ap = tau*ap
    ad = tau*ad
    x = x + ap*dx
    s = s + ad*ds
    y = y + ad*dy
    blas1_time += time()
  end

  blas1_time -= time()
  f = p'*x
  blas1_time += time()

  if print
    @printf "\nwo_inspector_hoisting_total_time = %f\n" time() - t0
    @printf "spgemm = %f fact = %f blas1 = %f trslv = %f spmv = %f\n" spgemm_time fact_time blas1_time trslv_time spmv_time
    @printf "iter %2i, resid = %9.2e, objval = %e\n" iter relResidual f[1][1]
  end

  x
end

# a reference implementation of interior-point method
# that uses Julia's default SpGEMM and Cholesky
function ipm_ref(A, b, p, print = true) # A: constraint coefficients, b: constraint rhs, p: objective
  t0 = time()
  (m,n) = size(A)

# set initial point, based on largest element in (A,b,p)
  bigM = maximum(A)
  bigM = maximum([norm(b, Inf) norm(p, Inf) bigM])
  x = 100*bigM*ones(n)
  s = x
  y = zeros(m)

  bc = 1 + maximum([norm(b) norm(p)])

  relResidual = NaN
  iter=1

  blas1_time = 0.
  spgemm_time = 0.
  fact_time = 0.
  trslv_time = 0.
  spmv_time = 0.

  D = speye_int32(n)

  for iter=1:200

    # compute residuals
    spmv_time -= time()
    Rd = A'*y + s - p
    Rp = A*x - b
    spmv_time += time()

    blas1_time -= time()
    Rc = x.*s
    mu = mean(Rc)
    relResidual = norm([Rd; Rp; Rc])/bc
    blas1_time += time()

    if print
      #@printf "iter %2i: log10(mu) = %5.2f, resid = %9.2e, obj = %9.2e\n" iter log10(mu) relResidual (p'*x)[1][1]
    end
    if (relResidual <= 1e-7 && mu <= 1e-7) break; end

    blas1_time -= time()
    Rc = Rc - min(0.1, 100*mu)*mu

    # set up the scaling matrix, and form the coefficient matrix for
    # the linear system
    d = min(5.e+15, x./s)
    blas1_time += time()

    spgemm_time -= time()
    D.nzval = d
    B = A*D*A'
    spgemm_time += time()

    # use the form of the Cholesky routine "cholinc" that's best
    # suited to interior-point methods
    fact_time -= time()
    R = cholfact(SparseMatrixCSC{Float64, Int64}(B))
    fact_time += time()

    # set up the right-hand side
    blas1_time -= time()
    t1 = x.*Rd - Rc;
    blas1_time += time()

    spmv_time -= time()
    t2 = -(Rp + A*(t1./s));
    spmv_time += time()

    # solve it and recover the other step components
    trslv_time -= time()
    dy = R\t2
    trslv_time += time()

    spmv_time -= time()
    temp = A'*dy
    spmv_time += time()

    blas1_time -= time()
    dx = (x.*temp + t1)./s
    ds = -(s.*dx + Rc)./x

    tau = max(.9995, 1 - mu)
    ap = -1/minimum([dx./x; -1])
    ad = -1/minimum([ds./s; -1])
    ap = tau*ap
    ad = tau*ad
    x = x + ap*dx
    s = s + ad*ds
    y = y + ad*dy
    blas1_time += time()
  end

  blas1_time -= time()
  f = p'*x
  blas1_time += time()

  if print
    @printf "\nref_total_time = %f\n" time() - t0
    @printf "spgemm = %f fact = %f blas1 = %f trslv = %f spmv = %f\n" spgemm_time fact_time blas1_time trslv_time spmv_time
    @printf "iter %2i, resid = %9.2e, objval = %e\n" iter relResidual f[1][1]
  end

  x
end

# min 2*x1 + x2 subject to x1 + x2 = 1, x1 >= 0, x2 >= 0
# expected solution: x1 = 0, x2 = 1, obj = 1
#A = sparse([1 1])
#b = [ 1 ]'
#p = [ 2 1 ]'
A = MatrixMarket2.mmread(string(ARGS[1], "-A.mtx"))'
b = vec(MatrixMarket.mmread(string(ARGS[1], "-b.mtx")))
p = vec(MatrixMarket.mmread(string(ARGS[1], "-p.mtx")))

m = size(A, 1)
n = size(A, 2)
println("Problem size = [$m $n]")

ipm_ref(A, b, p, false) # ignore timing of the first run
ipm_ref(A, b, p)

ipm_unopt(A, b, p, false)
ipm_unopt(A, b, p)

ipm(A, b, p, false)
ipm(A, b, p)

File: test/correctness/ipm/ipm.m
function [x,y,s,f] = pdip(Af) %,bf,pf)
% primal-dual interior-point method for problem % % min p'x s.t. Ax=b, x>=0, % % whose dual is % % max b'y  s.t. A'y+s=p,  s>=0.
%
% calling sequence:
%
% [x,y,s,f] = pdip(A,b,p)
%
% input: A is an m x n SPARSE constraint matrix.
%        b is an m x 1 right-hand side vector
%        p is an n x 1 cost vector.
%
% output: x is the  n x 1 solution of the primal problem
%         y is the m x 1 dual solution
%         s is the n x 1 vector of "dual slacks"
%         f is the optimal objective value

fprintf('%s\n', Af);
% min 2*x1 + x2 subject to x1 + x2 = 1, x1 >= 0, x2 >= 0
% expected solution: x1 = 0, x2 = 1, obj = 1
%A = sparse([1 1]);
%b = [ 1 ]';
%p = [ 2 1 ]';

% small example from http://cvxopt.org/examples/tutorial/lp.html
%A = sparse([ -1 1 1 0 0 ; -1 -1 0 1 0 ; 1 -2 0 0 1 ]);
%b = [ 1 -2 4 ]';
%p = [ 2 1 0 0 0 ]';

A = mmread([Af '-A.mtx']);
b = mmread([Af '-b.mtx']);
p = mmread([Af '-p.mtx']);
if size(p, 1) == 0
  p = ones(size(A,2),1);
end

fprintf('size(A): %d x %d\n', size(A));
if nargin ~= 1
  error('must have three input arguments'); end

if ~issparse(A)
  error('first input argument A must be a SPARSE matrix; possibly use sparse() to convert'); end

t0=cputime;
[m,n] = size(A);
if m <= 0 | n <= 0
  error('input matrix A must be nontrivial'); end

if n ~= length(p)
  error('size of vector p must match number of columns in A'); end
if m ~= length(b)
  error('size of vector b must match number of rows in A'); end

% set initial point, based on largest element in (A,b,p)
bigM = max(max(abs(A))); bigM = max([norm(b,inf), norm(p,inf), bigM]);
x = 100*bigM*ones(n,1); s = x;
y = zeros(m,1);

% find row/column ordering that gives a sparse Cholesky % factorization of ADA'
ordering = symamd(A*A');
norm(b)
norm(p)
size(p)
max(p)
bc = 1+max([norm(b), norm(p)])

for iter=1:200
  
% compute residuals
  Rd = A'*y+s-p;
  Rp = A*x-b;
  Rc = x.*s;
  mu = mean(Rc);
  relResidual = norm([Rd;Rp;Rc])/bc;
%  fprintf('iter %2i: mu = %9.2e, resid = %9.2e\n', iter, mu, relResidual);
  fprintf('iter %2i: log10(mu) = %5.2f, resid = %9.2e, obj = %9.2e\n', iter, log10(full(mu)), ...
          full(relResidual), p'*x);
  if(relResidual <= 1.e-7 & mu <= 1.e-7) break; end;
  Rc = Rc - min(0.1,100*mu)*mu;
  
  % set up the scaling matrix, and form the coefficient matrix for
  % the linear system
  d = min(5.e+15, x./s);
  B = A*sparse(1:n,1:n,d)*A';
  % use the form of the Cholesky routine "cholinc" that's best
  % suited to interior-point methods
  %R = cholinc(B(ordering,ordering),'inf'); % cholin obsolete in recent MATLAB versions
  R = chol(B(ordering, ordering));
  %R = sparse(ldl(full(B(ordering, ordering))))';
  %[L, U] = lu(full(B(ordering, ordering)));
  %R = sparse(diag(sqrt(abs(diag(U))))\U);
  
  % set up the right-hand side
  t1 = x.*Rd-Rc;
  t2 = -(Rp+A*(t1./s));
  
  % solve it and recover the other step components
  dy = zeros(m,1);
  dy(ordering) = R\(R'\t2(ordering));
  dx = (x.*(A'*dy)+t1)./s;
  ds = -(s.*dx+Rc)./x;
  
  tau = max(.9995,1-mu);
  ap = -1/min(min(dx./x),-1);
  ad = -1/min(min(ds./s),-1);
  ap = tau*ap;
  ad = tau*ad;
  x = x + ap*dx;
  s = s + ad*ds;
  y = y + ad*dy;
end

fprintf('iter %2i, resid = %9.2e\n', iter, relResidual)

f = p'*x;

% convert x,y,s to full data structures
x=full(x); s=full(s); y=full(y);

fprintf('Done!\t[m n] = [%g %g]\tCPU = %g\n', m, n, cputime-t0); fprintf('Optimal objective value (f) = %e\n', f); return;  

